etcd/v3/kv_etcd.go

```diff
@@ -491,47 +491,54 @@ func (et *etcdKV) CompareAndSet(
 		opts = append(opts, e.WithLease(leaseResult.ID))
 	}
 
-	ctx, cancel := et.Context()
+	cmp := e.Compare(e.Value(key), "=", string(prevValue))
 	if (flags & kvdb.KVModifiedIndex) != 0 {
+		cmp = e.Compare(e.ModRevision(key), "=", int64(kvp.ModifiedIndex))
+	}
+retry:
+	for {
+		ctx, cancel := et.Context()
 		txnResponse, txnErr = et.kvClient.Txn(ctx).
-			If(e.Compare(e.ModRevision(key), "=", int64(kvp.ModifiedIndex))).
+			If(cmp).
 			Then(e.OpPut(key, string(kvp.Value), opts...)).
 			Commit()
 		cancel()
 		if txnErr != nil {
-			return nil, txnErr
-		}
-		if txnResponse.Succeeded == false {
-			if len(txnResponse.Responses) == 0 {
-				logrus.Infof("Etcd did not return any transaction responses for key (%v)", kvp.Key)
-			} else {
-				for i, responseOp := range txnResponse.Responses {
-					logrus.Infof("Etcd transaction Response: %v %v", i, responseOp.String())
+			if strings.Contains(txnErr.Error(), rpctypes.ErrGRPCTimeout.Error()) {
+				// server timeout
+				kvPair, err := et.Get(kvp.Key)
+				if err != nil {
+					return nil, txnErr
+				}
+				if kvPair.ModifiedIndex == kvp.ModifiedIndex {
+					// update did not succeed, retry
+					continue retry
```

Comment:
Shouldn't we error out after fixed amount of retries instead of blocking the caller of the API ? Though there is a high chance that if the above Get succeeds we will not see a server timeout for the subsequent CompareAndSet.